skbio/sequence/_sequence.py

```diff
@@ -1408,6 +1408,74 @@ def count(self, subsequence, start=None, end=None):
         return self._string.count(
             self._munge_to_bytestring(subsequence, "count"), start, end)
 
+    @experimental(as_of="0.4.2-dev")
+    def replace(self, where, character):
+        """Replace values in this sequence with a different character.
+
+        Parameters
+        ----------
+        where : 1D array_like (bool) or iterable (slices or ints) or str
+            Indicates positions in the sequence to replace with `character`.
+            Can be a boolean vector, an iterable of indices/slices, or a
+            string that is a key in `positional_metadata` pointing to a
+            boolean vector.
+        character: str
+            Character that will replace chosen items in this sequence.
+
+        Returns
+        -------
+        Sequence
+            Copy of this sequence, with chosen items replaced with chosen
+            character. All metadata is retained.
+
+        Examples
+        --------
+        Let's create and display a Sequence:
+
+        >>> from skbio import Sequence
+        >>> sequence = Sequence('GGTACCAACG')
+        >>> str(sequence)
+        'GGTACCAACG'
+
+        Let's call ``replace`` on the Sequence using a boolean vector for
+        ``where`` and assign it to a new variable:
+
+        >>> seq = sequence.replace([False, False, False, True, False, False,
+        ...                         True, True, False, False], '-')
+
+        Let's take a look at the new Sequence:
+
+        >>> str(seq)
+        'GGT-CC--CG'
+
+        Other types of input are accepted by the ``where`` parameter. Let's
+        pass in a list of indices and slices instead and compare it to ``seq``:
+
+        >>> str(seq) == str(sequence.replace([3, slice(6, 8)], '-'))
+        True
+
+        ``where`` also accepts a boolean vector contained in
+        ``Sequence.positional_metadata``:
+
+        >>> sequence.positional_metadata = {'where':
+        ...                                 [False, False, False, True, False,
+        ...                                  False, True, True, False, False]}
+
+        Let's pass in the key ``'where'`` and compare to ``seq``:
+
+        >>> str(seq) == str(sequence.replace('where', '-'))
+        True
+
+        """
+        index = self._munge_to_index_array(where)
+        seq_bytes = self._bytes.copy()
+        if type(character) is not bytes:
+            character = character.encode('ascii')
+        seq_bytes[index] = ord(character)
+        return self.__class__(seq_bytes,
+                              metadata=self._metadata,
```

Comment:
This is a very minor/niche performance bug, but it would be better to use `has_metadata` and `has_positional_metadata` so that a sequence with "empty" metadata (e.g., `{}` and `pd.DataFrame()`) doesn't propagate its empty metadata to the new sequence (the new sequence should have its metadata set to `None` internally). This is hard to explain in words so here's the code I'd use and a unit test, let me know if you have any questions about it:

```python
metadata = None
if self.has_metadata():
    metadata = self.metadata
positional_metadata = None
if self.has_positional_metadata():
    positional_metadata = self.positional_metadata
return self.__class__(seq_bytes, metadata=metadata, positional_metadata=positional_metadata)

def test_replace_handles_empty_but_present_metadata_efficiently(self):
    seq = Sequence('ACGT', metadata={}, positional_metadata=pd.DataFrame(index=range(4)))
    obs = seq.replace([True, False, False, True], 'Z')
    self.assertIsNone(obs._metadata)
    self.assertIsNone(obs._positional_metadata)
```